packages/core/src/agent/MessageSender.ts

```diff
@@ -399,14 +356,29 @@ export class MessageSender {
 
     if (connection.theirDid) {
       this.logger.debug(`Resolving services for connection theirDid ${connection.theirDid}.`)
-      didCommServices = await this.retrieveServicesFromDid(connection.theirDid)
+      didCommServices = await this.didCommDocumentService.resolveServicesFromDid(connection.theirDid)
     } else if (outOfBand) {
-      this.logger.debug(`Resolving services from out-of-band record ${outOfBand?.id}.`)
+      this.logger.debug(`Resolving services from out-of-band record ${outOfBand.id}.`)
       if (connection.isRequester) {
-        for (const service of outOfBand.outOfBandInvitation.services) {
+        for (const service of outOfBand.outOfBandInvitation.getServices()) {
           // Resolve dids to DIDDocs to retrieve services
           if (typeof service === 'string') {
-            didCommServices = await this.retrieveServicesFromDid(service)
+            this.logger.debug(`Resolving services for did ${service}.`)
+            didCommServices = await this.didCommDocumentService.resolveServicesFromDid(service)
+
+            // store recipientKeyFingerprints in the oob record
+            const oldRecipientKeyFingerprints = outOfBand.getTag('recipientKeyFingerprints') as string[]
+            if (!oldRecipientKeyFingerprints?.length) {
+              const allRecipientKeys = didCommServices.reduce<Key[]>(
+                (aggr, { recipientKeys }) => [...aggr, ...recipientKeys],
+                []
+              )
+              outOfBand.setTag(
+                'recipientKeyFingerprints',
+                allRecipientKeys.map((key) => key.fingerprint)
+              )
+              await this.outOfBandRepository.update(outOfBand)
```

Comment:
Did you place this here to prevent needing to resolve the did(s) multiple times? If so we should probably add some priority to caching

It feels like this maybe better placed inside the createInvitation method. What do you think?

Reply:
yes, it feels unclean to put it here, but this is the actual place where the service did resolution is performed currently. Caching somewhere in the resolver might be a viable solution. But it is always tricky to say when the cache should get invalidated.